File: shared/src/main/java/org/openscreentime/shared/model/Streak.kt
package org.openscreentime.shared.model

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

/**
 * Consecutive recently-completed days (not including today, which isn't over
 * yet) that stayed under the currently-configured daily limit and unlock goal
 * (see #13). This is v1: it checks each past day against *today's* configured
 * limit/goal rather than reconstructing what was configured on that day, so a
 * recent limit change can make the streak read slightly generously or
 * conservatively for a few days - an acceptable simplification per the issue.
 *
 * A day with no synced data at all doesn't count as "under the limit" (we
 * can't confirm it), so it breaks the streak rather than being skipped.
 *
 * This is display-only positive reinforcement: it only ever counts up from
 * zero. There is deliberately no equivalent "days over limit" or "broken
 * streak" computation anywhere - see the design-principle check on #13.
 */
fun computeStreak(
    recentStats: List<DailyStats>,
    dailyLimitMinutes: Int,
    dailyUnlockGoal: Int?
): Int {
    val byDate = recentStats.associateBy { it.date }
    val limitMs = dailyLimitMinutes * 60_000L
    val calendar = Calendar.getInstance()
    val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    var streak = 0
    calendar.add(Calendar.DAY_OF_YEAR, -1)
    while (true) {
        val date = format.format(calendar.time)
        val stats = byDate[date] ?: break
        val underTime = stats.totalScreenTimeMs <= limitMs
        val underUnlocks = dailyUnlockGoal == null || stats.unlockCount <= dailyUnlockGoal
        if (!underTime || !underUnlocks) break
        streak++
        calendar.add(Calendar.DAY_OF_YEAR, -1)
    }
    return streak
}

File: ios/Sources/OpenScreenTimeParentKit/Models/Streak.swift
import Foundation

/// Mirrors `shared/model/Streak.kt` exactly - consecutive recently-completed days (not
/// including today) under the currently-configured daily limit and unlock goal (see #13).
/// Display-only positive reinforcement: only ever counts up from zero, no "broken streak"
/// equivalent - see the design-principle note on the Android side.
func computeStreak(recentStats: [DailyStats], dailyLimitMinutes: Int, dailyUnlockGoal: Int?) -> Int {
    let byDate = Dictionary(uniqueKeysWithValues: recentStats.map { ($0.date, $0) })
    let limitMs = Int64(dailyLimitMinutes) * 60_000
    let calendar = Calendar(identifier: .gregorian)
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd"
    formatter.locale = Locale(identifier: "en_US_POSIX")

    var streak = 0
    guard var cursor = calendar.date(byAdding: .day, value: -1, to: Date()) else { return 0 }
    while true {
        let dateStr = formatter.string(from: cursor)
        guard let stats = byDate[dateStr] else { break }
        let underTime = stats.totalScreenTimeMs <= limitMs
        let underUnlocks = dailyUnlockGoal == nil || stats.unlockCount <= dailyUnlockGoal!
        if !underTime || !underUnlocks { break }
        streak += 1
        // Never expected to fail for simple day arithmetic, but break rather than risk an
        // infinite loop (re-checking the same date forever) if it somehow ever does.
        guard let previous = calendar.date(byAdding: .day, value: -1, to: cursor) else { break }
        cursor = previous
    }
    return streak
}
